File: Scala/src/main/scala/TextbookTrack/Chapter06/BA6A.scala
package TextbookTrack.Chapter06

object BA6A {
  import scala.annotation.tailrec

  sealed trait Sign
  final case object Plus extends Sign {
    override def toString: String = "+"
  }
  final case object Minus extends Sign {
    override def toString: String = "-"
  }

  object Sign {
    def apply(char: Char): Sign = char match {
      case '+' => Plus
      case '-' => Minus
      case _ => throw new Exception("Unknow sign.")
    }
  }

  final case class SignedNumber(sign: Sign, number: Int) {
    override def toString: String = s"$sign$number"
  }

  final case class SignedPermutation(permutation: Vector[SignedNumber]) {
    override def toString: String = permutation.mkString("(", " ", ")")

    val length: Int = permutation.length

    def apply(ix: Int): SignedNumber = permutation(ix)

    def isKSorted(k: Int): Boolean = {
      val SignedNumber(sign, number) = permutation(k)
      sign == Plus && number == k + 1
    }

    def drop(k: Int): Vector[SignedNumber] = permutation.drop(k)

    def slice(a: Int, b: Int): Vector[SignedNumber] = permutation.slice(a, b)

    def take(k: Int): Vector[SignedNumber] = permutation.take(k)

    def zipWithIndex: Vector[(SignedNumber, Int)] = permutation.zipWithIndex
  }

  def readPermutation(line: String): SignedPermutation = {
    val permutations: Vector[SignedNumber] =
      line
        .dropRight(1)
        .drop(1)
        .split(" ")
        .map{ item => SignedNumber(Sign(item(0)), item.drop(1).toInt) }
        .toVector
    SignedPermutation(permutations)
  }

  private def findKInPermutationSuffix(permutation: SignedPermutation, k: Int): Option[Int] = {
    permutation
      .drop(k)
      .zipWithIndex
      .find{ case (SignedNumber(_, number), _) => number == k + 1 }
      .map{ case (_, ix) => ix + k }
  }

  private def performKReversal(permutation: SignedPermutation, k: Int): SignedPermutation = {
    findKInPermutationSuffix(permutation, k) match {
      case None => throw new Exception(s"The number $k was not found in permutation.")
      case Some(index) =>
        val kReversal: Vector[SignedNumber] =
          Iterator(permutation.take(k), permutation.slice(k, index + 1).reverse, permutation.drop(index + 1))
            .flatten
            .zipWithIndex
            .map{ case (item, ix) => if (k <= ix && ix <= index) oppositeSign(item) else item }
            .toVector
        SignedPermutation(kReversal)
    }
  }

  private def hasNegativeSign(permutation: SignedPermutation, k: Int): Boolean =
    permutation(k).sign == Minus

  private def oppositeSign(signedNumber: SignedNumber): SignedNumber =
    if (signedNumber.sign == Plus) signedNumber.copy(sign = Minus)
    else signedNumber.copy(sign = Plus)

  private def changeSign(permutation: SignedPermutation, k: Int): SignedPermutation =
    SignedPermutation(permutation.zipWithIndex.map{ case (item, ix) => if (ix == k) oppositeSign(item) else item })

  def greedySorting(permutation: SignedPermutation): List[SignedPermutation] = {
    @tailrec
    def loop(reversals: List[SignedPermutation],
             currentPermutation: SignedPermutation,
             k: Int): List[SignedPermutation] = {
      if (k == permutation.length) reversals.reverse
      else if (permutation.isKSorted(k)) loop(reversals, currentPermutation, k + 1)
      else {
        val nextPermutation: SignedPermutation = performKReversal(currentPermutation, k)
        if (hasNegativeSign(nextPermutation, k)) {
          val signCorrectedPermutation: SignedPermutation = changeSign(nextPermutation, k)
          loop(signCorrectedPermutation :: nextPermutation :: reversals, signCorrectedPermutation, k + 1)
        }
        else loop(nextPermutation :: reversals, nextPermutation, k + 1)
      }
    }

    loop(Nil, permutation, 0)
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val permutation: SignedPermutation = readPermutation(reader.next())
    val result: List[SignedPermutation] = greedySorting(permutation)
    result.foreach(println)
  }
}

File: Python/textbook_track/chapter06/ba6a.py
# Find the Minimum Number of Coins Needed to Make Change
import itertools as it
import sys

from enum import Enum


class Sign(Enum):
    PLUS = '+'
    MINUS = '-'


class SignedNumber:
    def __init__(self, sign, number):
        self._sign = sign
        self._number = number

    def __repr__(self):
        return f'{self._sign.value}{self._number}'

    @property
    def sign(self):
        return self._sign

    @property
    def number(self):
        return self._number


class SignedPermutation:
    def __init__(self, permutation):
        self._permutation = tuple(permutation)

    def __repr__(self):
        return f"({' '.join(map(str, self._permutation))})"

    def __len__(self):
        return len(self._permutation)

    def __getitem__(self, k):
        return self._permutation[k]

    def __iter__(self):
        yield from self._permutation


def read_permutation(line):
    line = line.replace('(', '').replace(')', '')
    permutation = []
    for item in line.split():
        signed_number = SignedNumber(Sign(item[0]), int(item[1:]))
        permutation.append(signed_number)
    return SignedPermutation(permutation)


def find_k_in_permutation_suffix(permutation, k):
    for ix, item in enumerate(permutation[k:], start=k):
        if item.number == k + 1:
            return ix
    return None


def perform_k_reversal(permutation, k):
    index = find_k_in_permutation_suffix(permutation, k)
    k_reversal = []
    for ix, item in enumerate(it.chain(permutation[:k], reversed(permutation[k:index+1]), permutation[index+1:])):
        if k <= ix <= index:
            signed_number = opposite_sign(item)
        else:
            signed_number = item
        k_reversal.append(signed_number)

    return SignedPermutation(k_reversal)


def is_k_sorted(permutation, k):
    signed_number = permutation[k]
    return signed_number.sign == Sign.PLUS and signed_number.number == k + 1


def has_negative_sign(permutation, k):
    return permutation[k].sign == Sign.MINUS


def opposite_sign(signed_number):
    if signed_number.sign == Sign.PLUS:
        return SignedNumber(Sign.MINUS, signed_number.number)

    return SignedNumber(Sign.PLUS, signed_number.number)


def change_sign(permutation, k):
    updated_permutation = [opposite_sign(item) if ix == k else item for ix, item in enumerate(permutation)]
    return SignedPermutation(updated_permutation)


def greedy_sorting(permutation):
    n = len(permutation)
    reversals = []
    for k in range(n):
        if not is_k_sorted(permutation, k):
            permutation = perform_k_reversal(permutation, k)
            reversals.append(permutation)
            if has_negative_sign(permutation, k):
                permutation = change_sign(permutation, k)
                reversals.append(permutation)
    return reversals


def main():
    reader = sys.stdin
    permutation = read_permutation(next(reader))
    result = greedy_sorting(permutation)
    for reversal in result:
        print(reversal)


if __name__ == '__main__':
    main()
